chainindex/indexer.go

```diff
@@ -0,0 +1,475 @@
+	revertTipsetStmt                      *sql.Stmt
+	getMaxNonRevertedTipsetStmt           *sql.Stmt
+	tipsetExistsStmt                      *sql.Stmt
+	tipsetUnRevertStmt                    *sql.Stmt
+	removeRevertedTipsetsBeforeHeightStmt *sql.Stmt
+	removeTipsetsBeforeHeightStmt         *sql.Stmt
+	deleteEthHashesOlderThanStmt          *sql.Stmt
+	revertTipsetsFromHeightStmt           *sql.Stmt
+	countMessagesStmt                     *sql.Stmt
+	minNonRevertedHeightStmt              *sql.Stmt
+	tipsetExistsNotRevertedStmt           *sql.Stmt
+
+	gcRetentionEpochs int64
+
+	mu           sync.Mutex
+	updateSubs   map[uint64]*updateSub
+	subIdCounter uint64
+
+	closeLk sync.RWMutex
+	closed  bool
+}
+
+func NewSqliteIndexer(path string, cs ChainStore, gcRetentionEpochs int64) (si *SqliteIndexer, err error) {
+	db, _, err := sqlite.Open(path)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to setup message index db: %w", err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	defer func() {
+		if err != nil {
+			_ = db.Close()
+			cancel()
+		}
+	}()
+
+	err = sqlite.InitDb(ctx, "chain index", db, ddls, []sqlite.MigrationFunc{})
+	if err != nil {
+		return nil, xerrors.Errorf("failed to init message index db: %w", err)
+	}
+
+	si = &SqliteIndexer{
+		ctx:               ctx,
+		cancel:            cancel,
+		db:                db,
+		cs:                cs,
+		updateSubs:        make(map[uint64]*updateSub),
+		subIdCounter:      0,
+		gcRetentionEpochs: gcRetentionEpochs,
+	}
+	if err = si.prepareStatements(); err != nil {
+		return nil, xerrors.Errorf("failed to prepare statements: %w", err)
+	}
+
+	si.wg.Add(1)
+	go si.gcLoop()
+
+	return si, nil
+}
+
+// ReconcileWithChain ensures that the index is consistent with the current chain state.
+// It performs the following steps:
+//  1. Checks if the index is empty. If so, it returns immediately as there's nothing to reconcile.
+//  2. Finds the lowest non-reverted height in the index.
+//  3. Walks backwards from the current chain head until it finds a tipset that exists
+//     in the index and is not marked as reverted.
+//  4. Sets a boundary epoch just above this found tipset.
+//  5. Marks all tipsets above this boundary as reverted, ensuring consistency with the current chain state.
+//  6. Applies all missing un-indexed tipsets starting from the last matching tipset b/w index and canonical chain
+//     to the current chain head.
+//
+// This function is crucial for maintaining index integrity, especially after chain reorgs.
+// It ensures that the index accurately reflects the current state of the blockchain.
+func (si *SqliteIndexer) ReconcileWithChain(ctx context.Context, currHead *types.TipSet) error {
+	si.closeLk.RLock()
+	if si.closed {
+		return ErrClosed
+	}
+	si.closeLk.RUnlock()
+
+	if currHead == nil {
+		return nil
+	}
+
+	return withTx(ctx, si.db, func(tx *sql.Tx) error {
+		row := tx.StmtContext(ctx, si.countMessagesStmt).QueryRowContext(ctx)
```

Comment:
this seems unnecessary and could be quite slow on a large db, could we use the next call to determine whether to proceed or not? if it returns no results then can we assume there's no rows? I don't remember if there's a special error type for that or not but perhaps we could even differentiate between "no results" and "other error".

Reply:
Made this efficient using a simpler query.